cpp/daal/src/algorithms/objective_function/cross_entropy_loss/cross_entropy_loss_dense_default_batch_impl.i

```diff
@@ -113,17 +111,49 @@ void CrossEntropyLossKernel<algorithmFPType, method, cpu>::softmax(const algorit
         }
     }
     daal::internal::Math<algorithmFPType, cpu>::vExp(nRows * nCols, res, res);
-    for (size_t iRow = 0; iRow < nRows; ++iRow)
+    if (softmaxSums != nullptr && yLocal != nullptr)
     {
-        algorithmFPType * pRes = res + iRow * nCols;
-        algorithmFPType sum(0.);
-        PRAGMA_IVDEP
-        PRAGMA_VECTOR_ALWAYS
-        for (size_t i = 0; i < nCols; ++i) sum += pRes[i];
-        sum = algorithmFPType(1.) / sum;
-        PRAGMA_IVDEP
-        PRAGMA_VECTOR_ALWAYS
-        for (size_t i = 0; i < nCols; ++i) pRes[i] *= sum;
+        for (size_t iRow = 0; iRow < nRows; ++iRow)
+        {
+            algorithmFPType * const pRes = res + iRow * nCols;
+            algorithmFPType sum(0.);
+            PRAGMA_IVDEP
+            PRAGMA_VECTOR_ALWAYS
+            for (size_t i = 0; i < nCols; ++i)
+            {
+                sum += pRes[i];
+            }
+            sum = static_cast<algorithmFPType>(1.) / sum;
+            PRAGMA_IVDEP
+            PRAGMA_VECTOR_ALWAYS
+            for (size_t i = 0; i < nCols; ++i)
+            {
+                pRes[i] *= sum;
+                softmaxSums[i] += pRes[i];
+            }
+            softmaxSums[static_cast<size_t>(yLocal[iRow])] -= 1;
+        }
+    }
+    else
+    {
+        for (size_t iRow = 0; iRow < nRows; ++iRow)
+        {
+            algorithmFPType * const pRes = res + iRow * nCols;
+            algorithmFPType sum(0.);
+            PRAGMA_IVDEP
+            PRAGMA_VECTOR_ALWAYS
+            for (size_t i = 0; i < nCols; ++i)
+            {
+                sum += pRes[i];
+            }
+            sum = static_cast<algorithmFPType>(1.) / sum;
+            PRAGMA_IVDEP
+            PRAGMA_VECTOR_ALWAYS
+            for (size_t i = 0; i < nCols; ++i)
+            {
+                pRes[i] *= sum;
```

Comment:
`axpy` can be applied if this branch is applicable still

Reply:
I think that we merge this code "as is", but in the future do some explorations about perf gains of using axpy here.